plugin/inline_completion.py

```diff
@@ -0,0 +1,215 @@
+    }}
+</style>
+<body id="lsp-inline-completion">
+    <div class="completion-content">{content}</div>
+</body>"""
+
+
+class InlineCompletionData:
+
+    def __init__(self, view: sublime.View, key: str) -> None:
+        self.visible = False
+        self.index = 0
+        self.position = 0
+        self.items: list[tuple[str, sublime.Region, str, Optional[Command]]] = []
+        self._view = view
+        self._phantom_set = sublime.PhantomSet(view, key)
+
+    def render_async(self, index: int) -> None:
+        style = self._view.style_for_scope('comment meta.inline-completion.lsp')
+        color = style['foreground']
+        font_style = 'italic' if style['italic'] else 'normal'
+        font_weight = 'bold' if style['bold'] else 'normal'
+        region = sublime.Region(self.position)
+        item = self.items[index]
+        first_line, *more_lines = item[2][len(item[1]):].splitlines()
+        phantoms = [sublime.Phantom(
+            region,
+            PHANTOM_HTML.format(
+                color=color,
+                font_style=font_style,
+                font_weight=font_weight,
+                content=self._normalize_html(first_line)
+            ),
+            sublime.PhantomLayout.INLINE
+        )]
+        if more_lines:
+            phantoms.append(
+                sublime.Phantom(
+                    region,
+                    PHANTOM_HTML.format(
+                        color=color,
+                        font_style=font_style,
+                        font_weight=font_weight,
+                        content='<br>'.join(self._normalize_html(line) for line in more_lines)
+                    ),
+                    sublime.PhantomLayout.BLOCK
+                )
+            )
+        sublime.set_timeout(lambda: self._render(phantoms, index))
+        self.visible = True
+
+    def _render(self, phantoms: list[sublime.Phantom], index: int) -> None:
+        self.index = index
+        self._phantom_set.update(phantoms)
+
+    def clear_async(self) -> None:
+        if self.visible:
+            sublime.set_timeout(self._clear)
+            self.visible = False
+
+    def _clear(self) -> None:
+        self._phantom_set.update([])
+
+    def _normalize_html(self, content: str) -> str:
+        return html.escape(content).replace(' ', '&nbsp;')
+
+
+class LspInlineCompletionCommand(LspTextCommand):
+
+    capability = 'inlineCompletionProvider'
+
+    def run(self, edit: sublime.Edit, event: dict | None = None, point: int | None = None) -> None:
+        sublime.set_timeout_async(partial(self._run_async, event, point))
+
+    def _run_async(self, event: dict | None = None, point: int | None = None) -> None:
+        position = get_position(self.view, event, point)
+        if position is None:
+            return
+        session = self.best_session(self.capability, point)
+        if not session:
+            return
+        if self.view.settings().get('mini_auto_complete', False):
```

Comment:
hahahaha I was wondering why this PR doesn't work for me, it was due to this check.
I had `mini_auto_complete` enabled.

Reply:
I am thinking if this should be documented.

Reply:
this was addressed with [3063b62](https://github.com/sublimelsp/LSP/pull/2552/commits/3063b628d698339d3641d029caf64fbb2bd6d5dd)